File: leetcode981/leetcode981_my_version.cpp
class TimeMap {
public:
    /** Initialize your data structure here. */
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        hashmap[key].push_back({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        string ans = "";
        
        auto& arr = hashmap[key];
        
        int start = 0, end = arr.size();
        
        while (start < end) {
            int mid = start + (end - start) / 2;
            
            if (arr[mid].first < timestamp) {
                ans = arr[mid].second;
                start = mid + 1;
            }
            else if (arr[mid].first > timestamp) {
                end = mid;
            }
            else {
                return arr[mid].second;
            }
        }
        
        return ans;
    }
    
private:
    unordered_map<string, vector<pair<int, string>>> hashmap;
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */


// Runtime: 388 ms, faster than 84.90% of C++ online submissions for Time Based Key-Value Store.
// Memory Usage: 134.7 MB, less than 100.00% of C++ online submissions for Time Based Key-Value Store.
// time complexity: O(1) set, O(log n) get
// space complexity: O(1)

File: leetcode981/leetcode981_my_version.java
class TimeMap {
    Map<String, ArrayList<ArrayList<String>>> map;

    /** Initialize your data structure here. */
    public TimeMap() {
        map = new HashMap<String, ArrayList<ArrayList<String>>>();
    }

    public void set(String key, String value, int timestamp) {
        if (map.get(key) == null) {
            map.put(key, new ArrayList<ArrayList<String>>());
        }

        map.get(key).add(new ArrayList<String>(List.of(value, String.valueOf(timestamp))));
    }

    public String get(String key, int timestamp) {
        // String ans = "";
        // int t;
        // for (ArrayList<String> arrL : map.get(key))
        // {
        // if (Integer.valueOf(arrL.get(1)) <= timestamp)
        // {
        // t = Integer.valueOf(arrL.get(1));
        // ans = arrL.get(0);
        // }
        // else
        // {
        // break;
        // }
        // }

        // return ans;

        String ansStr = "";
        ArrayList<ArrayList<String>> ans = map.get(key);

        // binary search
        int left = 0, right = ans.size();

        while (left < right) {
            int mid = left + (right - left) / 2;
            int t = Integer.valueOf(ans.get(mid).get(1));
            if (t == timestamp) {
                return ans.get(mid).get(0);
            } else if (t < timestamp) {
                ansStr = ans.get(mid).get(0);
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return ansStr;
    }
}

/**
 * Your TimeMap object will be instantiated and called as such: TimeMap obj =
 * new TimeMap(); obj.set(key,value,timestamp); String param_2 =
 * obj.get(key,timestamp);
 */